File: hw-7B/C++/B-customs.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <utility>

using namespace std;

bool cmp(pair<int, int> p1, pair<int, int> p2){
       
    if (p1.first == p2.first)
        return p1.second < p2.second;
    return (p1.first < p2.first);
}

int arrivedGoods(vector <pair<int, int>> events){

    sort(events.begin(), events.end(), cmp);
    int arrived = 0, maxArrived = 0;
    for (int i=0; i<events.size(); i++){
        if (events[i].second == 1){
            arrived++;
        }
        else {
            arrived--;
        }
        maxArrived = max(maxArrived, arrived);
    }
    return maxArrived; 
}

int main(){

    // Define a minimum of machines required for preprocessing all goods after their arrival

    int n, x, y;
    cin >> n;
    int start = 1, end = -1;
    vector <pair<int, int>> events;
    for (int i=0; i<n; i++){
        cin >> x >> y;
        events.push_back(make_pair(x, start));
        events.push_back(make_pair(x + y, end));
    }
    int res = arrivedGoods(events);
    cout << res << "\n";
    return 0;
}

File: hw-7B/Python/B-customs.py
def arrived_goods(events):

    events.sort()
    print(events)
    arrived = 0
    max_arrived = 0
    for event in events:
        if event[1] == 1:
            arrived += 1
        elif event[1] == -1:
            arrived -= 1
        max_arrived = max(max_arrived, arrived)
    return max_arrived


if __name__ == '__main__':

    """Define the minimum of machines required for preprocessing all goods after their arrival"""

    n = int(input())
    events = []
    end, start = -1, 1  # out before in
    for i in range(n):
        a, b = map(int, input().split())
        events.append((a, start))
        events.append((a + b, end))
    res = arrived_goods(events)
    print(res)

